### PR TITLE
fix(mlx): decouple tokenizer loading from model configs

### PR DESCRIPTION
**Pull request by Lyxot**
Transformers 5.5 can fail to load an otherwise usable tokenizer because `AutoTokenizer.from_pretrained` constructs the model config first. Published configs for Phi-4 multimodal, EXAONE, DeepSeek V3.2, Step 3.5, LongCat Flash Ngram, Nemotron NAS/H, MiniCPM3 and DBRX trigger unsupported-model or config-validation errors before tokenization.

Load serialized fast tokenizers from their tokenizer metadata and files, preserving chat templates and special/added tokens. Declared remote tokenizer classes use a neutral config so their selection also avoids model validation. Keep mlx-lm's tokenizer wrapper, detokenizer selection and EOS handling intact.

Propagate the caller's `trust_remote_code` choice, defaulting to `False`, through ordinary and distributed text loads, nested VLM tokenizer loads, processor recovery and image-processor recovery. Prefer a serialized fast tokenizer for tiktoken wrappers when available. Remote-only refusals name the Python file and opt-in flag, including refusals swallowed by mlx-vlm's processor dispatch. A successful native processor fallback still takes precedence. Nested tokenizer dispatch is scoped to the loading thread, serialized with a reentrant lock, and restored on exit. Studio already forwards the flag and requires no change.

### Validation

- 29 focused loader regressions passed on transformers 5.5.0 after rebasing onto current `main`; 11 tokenizer/processor regressions passed on 4.51.3 with compatible dependencies.
- 14 isolated mutation checks failed their intended test selections, covering config resolution, fast-tokenizer preference, trust propagation in both directions, distributed loading, native fallback precedence and refusal preservation.
- Tested published tokenizer/config files with tiny random checkpoints and no substitute tokenizers or downloaded model weights. All nine config-failure families passed tokenizer and model loading; eight generated. Of 18 remote-code families, 17 passed tokenizer/processor loading with explicit trust and 16 completed model loading and generation.

```bash
python -m pytest tests/test_mlx_save_export_regressions.py tests/test_mlx_distributed_loader.py -q -k 'tokenizer or processor or distributed or strict_fallback'
```

### Scope and remaining failures

The [published DBRX repo](https://huggingface.co/mlx-community/dbrx-instruct-4bit/tree/main) has no `tokenizer.json`, so its `tiktoken.py` wrapper still requires explicit trust. Native Kimi K3 and Plamo2VL processors avoid repository code and remain usable without that opt-in.

This does not claim complete family support: Nemotron NAS reaches an mlx-lm cache failure after tokenization; Plamo2VL reaches an existing layer-indexing fixup failure; Youtu VL's published remote image processor imports the removed `DefaultFastImageProcessorKwargs`. Direct mlx-vlm reproductions also retain ERNIE's missing `.vocab` detokenizer failure and VoiceChat's missing generic processor. Backend model-config parsing and post-tokenizer model behavior are unchanged; no dependency source patches or family-specific shims are included.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
The `Repo tests (CPU)` failure is in the new test, not in your change. The production hunk is fine.

`transformers` 5.x hands out `AutoImageProcessor` as a torchvision-gated placeholder when torchvision is absent: its metaclass runs `requires_backends()` on every attribute access. `monkeypatch.setattr` reads the old value first, so

```python
monkeypatch.setattr(transformers.AutoImageProcessor, "from_pretrained", staticmethod(from_pretrained))
```

raises on that read and the test dies before reaching a line of `_build_vlm_image_processor_from_config`. The `Repo tests (CPU)` lane installs no torchvision on purpose (`consolidated-tests-ci.yml`), which is also what a Mac user installing `zoo[mlx]` gets, so this is the configuration the MLX path most needs to work in.

The empty `ImportError:` in the summary is upstream's doing: `BACKENDS_MAPPING["torchvision"]`'s template starts with a newline, so `raise ImportError("".join(failed))` puts the whole message on the following lines.

Reproduced locally against the CI pins (transformers 5.5.0, torch 2.10.0+cpu, python 3.12, no torchvision): `2 failed, 205 passed`, byte-identical traceback. Installing torchvision alone flips both to green, so torchvision absence is the only cause.

Patching the name the builder imports, rather than an attribute of the placeholder, works in both configurations:

```python
class FakeAutoImageProcessor:
    @staticmethod
    def from_pretrained(path, **kwargs):
        calls.append(kwargs["trust_remote_code"])
        return processor

# Replace the name the builder imports, never an attribute OF the real class:
# transformers 5.x hands out AutoImageProcessor as a torchvision-gated placeholder
# and every attribute access on it raises ImportError. torchvision is not a
# dependency of this package, so setattr on the class itself fails on every
# machine the MLX path actually supports.
monkeypatch.setattr(transformers, "AutoImageProcessor", FakeAutoImageProcessor)
```

`from transformers import AutoImageProcessor` reads the module `__dict__` first, so the fake is what the builder sees, and monkeypatch restores the placeholder on teardown. With that change the file is `208 passed` without torchvision and still green with it.

Please avoid `pytest.importorskip("torchvision")` here: it would turn CI green by deleting the trust-forwarding assertion in the only configuration that runs it.

Worth adding alongside it, since the current call already sits inside `try/except Exception: return None` and nothing pins that:

```python
def test_image_processor_builder_returns_none_without_torchvision(monkeypatch, tmp_path):
    """The caller treats None as "no image processor", so a gated AutoImageProcessor
    has to degrade instead of raising out of the loader."""
```

That way, if the `except Exception` wrapper is ever removed, it fails here with a legible assertion instead of surfacing as a VLM load crash on someone's Mac.

**Comment by danielhanchen** [comment]
Separately from the `Repo tests (CPU)` failure above, the `Core drift (3 upstream pins)` failure here is stale and needs a different remedy.

It fails on the `moe_utils` `sys.modules` leak from `test_compiler_dynamic_exec.py`, which #1169 and #1172 fixed in `tests/conftest.py`. `Core drift` is green on main at `9bbb63ce` and `5335e91f`.

A rerun does not clear it: `git merge-base --is-ancestor efb57a32 refs/pull/1170/merge` is false, because a rerun replays the merge commit as it was resolved when the run was first created, and this job executes the PR's own checked-out `tests/conftest.py`. Merging or rebasing main into the branch is what fixes it, and it will also pick up #1175, so do that before re-running anything here.

**Comment by danielhanchen** [comment]
Confirmed the loader path where AutoTokenizer.from_pretrained builds the model config first and dies before tokenization on transformers 5.5, so this is worth having. One thing before I review: the has_fast_file branch prefers the serialized fast tokenizer over a declared tiktoken wrapper even when trust_remote_code=True, could you confirm the resulting special-token and encoding behaviour matches the wrapper for a repo that ships both?

**Comment by danielhanchen** [comment]
Reproduced this end to end and pushed two commits to the branch. Summary of what I ran and what it found.

## The bug is real, and it is specific to transformers 5.x

I fetched the published metadata (config + tokenizer files only, no weights) for the families you name and ran each directory through `AutoTokenizer.from_pretrained` and then through your `_load_mlx_tokenizer`, on transformers 5.5.0 / torch 2.10.0+cpu:

| repo | before | after |
|---|---|---|
| microsoft/Phi-4-multimodal-instruct | `AttributeError: 'PreTrainedConfig' object has no attribute 'max_position_embeddings'` | loads |
| nvidia/Llama-3_1-Nemotron-51B-Instruct | `AttributeError: ... max_position_embeddings` | loads |
| deepseek-ai/DeepSeek-V3.2-Exp | `AttributeError: ... max_position_embeddings` | loads |
| nvidia/Nemotron-H-8B-Base-8K | `KeyError: '-'` | loads |
| openbmb/MiniCPM3-4B | `KeyError: "Missing required keys in rope_parameters for 'rope_type'='longrope': {'rope_theta'}"` | loads |
| LGAI-EXAONE/EXAONE-4.0-32B, stepfun-ai/step3, meituan-longcat/LongCat-Flash-Chat | both load | both load, identical |
| mlx-community/dbrx-instruct-4bit | `StrictDataclassFieldValidationError` on `moe_jitter_eps` | named refusal, as you documented |

The mechanism is worth writing down because it is narrower than "the config is built first". In transformers 5.5 `AutoTokenizer.from_pretrained` does always resolve the config first, but inside `except (ValueError, OSError)` with a `PreTrainedConfig` fallback. So a config that raises `ValueError` is harmless. The five that break all raise `AttributeError` or `KeyError`, which that handler does not catch. My first attempt at a repro used a synthetic unsupported `model_type` and passed cleanly, which is exactly why I went and got the real files.

On transformers 4.57.6 all twelve repos load identically before and after, so this is a 5.x-only fix and a no-op on the old pin.

## One regression, now fixed

`openai-community/gpt2` was the only control repo that changed behaviour:

```
BEFORE  class=GPT2Tokenizer     eos='<|endoftext|>'  bos='<|endoftext|>'  eos_token_id=50256
AFTER   class=TokenizersBackend eos=None             bos=None             eos_token_id=None
```

gpt2 declares no `tokenizer_class` and ships no `special_tokens_map.json`; its bos/eos/unk live in the tokenizer class `__init__` defaults. `has_fast_file` was therefore true with `class_name` None, so the branch loaded a bare `PreTrainedTokenizerFast` and the special tokens went with it. Encoding was unchanged, which is what made it silent, but an `eos_token` of None leaves mlx-lm generation with no stop token.

Blast radius across 23 real repos: 5 affected, all in the GPT-2 family (`gpt2`, `gpt2-medium`, `gpt2-large`, `gpt2-xl`, `distilgpt2`). Every other repo I checked declares a `tokenizer_class`.

Fixed in `bda8c42e` by resolving the class from `config.json`'s `model_type` read as a plain string and mapped through `TOKENIZER_MAPPING_NAMES`. Nothing instantiates a config object, so the validation failures this branch exists to dodge stay dodged; there is a test asserting that. After the fix all three controls and all five fixed families come back `identical`, with zero regressions.

## The two red checks

`Repo tests (CPU)` was your test, not your production hunk, and I reproduced it byte for byte locally (`2 failed, 205 passed`). `monkeypatch.setattr(transformers.AutoImageProcessor, "from_pretrained", ...)` reads the old value first, and without torchvision transformers 5.x hands out `AutoImageProcessor` as a gated placeholder that raises on every attribute read. Patching the name the builder imports works in both configurations. I also added the degrade-to-None test, since the call already sits inside `except Exception: return None` and nothing pinned that.

`Core drift` was stale, from the `moe_utils` `sys.modules` leak fixed on main. Merged main in `222d9f63`, which also picks up #1175.

## Verification

`tests/test_mlx_save_export_regressions.py` + `tests/test_mlx_distributed_loader.py`, **223 passed** in each of three environments: transformers 5.5.0 without torchvision, transformers 5.5.0 with torchvision, transformers 4.57.6. Plus the neighbouring MLX suites (`save_export_edge_cases`, `module_exports`, `text_path_contract`, `preference`, `trainer_internals`, `autodetect_template_source`, `extra_special_tokens_compat`): 472 passed, 9 skipped.

32 edge cases around the new machinery all pass: refusal messaging with and without trust, scope restoration on clean exit / after an exception / nested, thread isolation of the `active` flag, 4 threads x 15 concurrent scopes leaving `AutoTokenizer.from_pretrained` pristine, bare hub ids not tripping the `Path(...).is_file()` probe, and malformed `auto_map` shapes (bare string, list, all-null, an integer) plus unparseable `config.json` degrading rather than raising.

Apple Silicon and Windows are running now on a staging repo, since the `MLX install + import smoke (Apple Silicon)` job here is label-gated and skips on fork PRs. I will post those results when they land.

## Still open, from my earlier question

The `has_fast_file` branch preferring the serialized tokenizer over a declared tiktoken wrapper: I could only test this on a synthetic repo, because the DBRX checkpoint ships no `tokenizer.json` and so never reaches that branch. On the synthetic case encoding is identical and no remote code is imported. If you know of a published repo that ships both, that is the one case still worth a real check.

**Comment by danielhanchen** [comment]
Answering my own earlier question, so it does not sit open on you.

I asked whether the `has_fast_file` branch preferring a serialized fast tokenizer over a declared tiktoken wrapper, even under `trust_remote_code=True`, could change special-token or encoding behaviour. It cannot, because the two never co-occur. Scanning the published tiktoken-wrapper repos for the combination the clause needs (a `tokenizer.json` present *and* `tiktoken` named in `tokenizer_class` / `auto_map`):

| repo | tokenizer.json | tiktoken in auto_map | clause fires |
|---|---|---|---|
| mlx-community/dbrx-instruct-4bit | no | yes | no |
| moonshotai/Kimi-K2-Instruct | no | yes | no |
| moonshotai/Moonlight-16B-A3B-Instruct | no | yes | no |
| mlx-community/Kimi-K2-Instruct-4bit | no | yes | no |
| Qwen-7B/14B/VL-Chat, chatglm3-6b, internlm2, Baichuan2, Yi-34B | no (Yi: yes, but no tiktoken) | no | no |

0 of 11. That is not a coincidence: a repo ships a tiktoken wrapper *because* it has no serialized fast tokenizer, so the clause is defensive rather than load-bearing. Nothing to compare and nothing to change. Metadata reads only, no repository code executed.

The `not remote` half of the same condition is the half that actually carries traffic, and that is where the gpt2 special-token loss I fixed in `bda8c42e` was hiding.

**Comment by danielhanchen** [comment]
Third commit, `d3ee66d0`, plus the compatibility work behind it.

## Forwards compatibility: the new imports were unpinned

`_load_mlx_tokenizer` reaches into `transformers.models.auto.tokenization_auto` for `get_tokenizer_config` and `tokenizer_class_from_name`, and my `_tokenizer_class_for_model_type` adds `TOKENIZER_MAPPING_NAMES`. None of the three is public API, and `tests/test_upstream_pinned_symbols_transformers.py` did not cover that module at all, so an upstream rename would have surfaced as an `ImportError` on a user's Mac rather than as a red `Core drift` lane. Added a pinned-symbol test in the file's existing style; it passes on v4.57.6, v5.0.0, v5.1.0, v5.2.0, v5.3.0, v5.5.0 and `main`.

## Backwards compatibility

| environment | result |
|---|---|
| transformers 5.5.0, no torchvision (the `Repo tests (CPU)` shape, and what `zoo[mlx]` gives a Mac user) | 223 passed |
| transformers 5.5.0 + torchvision <0.26 | 223 passed |
| transformers 4.57.6 (the lower-bound `Core drift` lane) | 223 passed |
| unsloth 2026.8.7 (old install) + this zoo | 223 passed, 32/32 edge cases |

On 4.57.6 all twelve repos load identically before and after, so the change is inert on the old pin. The old-install lane matters because it is the realistic upgrade path: a user on an older `unsloth` who bumps `unsloth_zoo` gets identical loader behaviour.

## Reachability, rather than an assertion that this is Mac-only

`unsloth_zoo/mlx/loader.py` imports `mlx.core` at module scope and MLX ships wheels only for darwin/arm64, so importing it anywhere else is a `ModuleNotFoundError`. Importing `unsloth_zoo` normally does not pull it in: the package registers `unsloth_zoo.mlx_loader` as a `_LazyMLXAlias` (no `__file__`), and touching `unsloth_zoo.mlx_loader.FastMLXModel` on Linux raises `AttributeError` without importing the changed module. All six symbols this PR adds or changes are confined to `unsloth_zoo/mlx/`, and the three mlx requirements carry `sys_platform == 'darwin' and platform_machine == 'arm64'` markers.

So of [Windows, Linux, WSL, Mac] x [NVIDIA, AMD, CPU], nine cells cannot execute a line of this diff, and the accelerator axis is degenerate on the remaining three. No browser-facing file is touched either; the diff is two test files and one loader.

Apple Silicon and Windows CI are still pending. The first staging account ran out of Actions capacity (13 runs queued, nothing started in an hour), so I moved both runs to one with live runners and will post the per-OS table when they land.

**Comment by danielhanchen** [comment]
Narrowed this to the config-decoupling fix in `6e2f50e3`. Apologies for editing your branch directly rather than asking first; the reasoning is below and every removal is one commit away from being restored.

## Why

The PR carried three separable changes:

1. decoupling tokenizer loading from model config resolution
2. propagating `trust_remote_code` through the VLM processor and image-processor paths
3. a refusal error naming the `.py` file a repo needs

Only the first fixes a broken load, and I have the measurements to say so rather than just assert it. Against published metadata for every family in your description, on transformers 5.5.0:

| family | before | after (decoupling alone) |
|---|---|---|
| Phi-4 multimodal, Nemotron NAS, DeepSeek V3.2 | `AttributeError: 'PreTrainedConfig' object has no attribute 'max_position_embeddings'` | loads |
| Nemotron H | `KeyError: '-'` | loads |
| MiniCPM3 | `KeyError: rope_parameters/longrope` | loads |

All five are fixed by (1) on its own. (3) changes no outcome for any repo I could find: DBRX fails before and after regardless, and dropping the custom refusal just surfaces transformers' own message, which already names the repository and the `trust_remote_code` opt-in. (2) has no case behind it where a load succeeds that would otherwise fail.

I also dropped the tiktoken clause in the fast-file branch. Of eleven published tiktoken-wrapper repos (DBRX, Kimi K2, Moonlight, the mlx-community conversions), **none** ships a `tokenizer.json` -- shipping one is precisely why a repo would not need the wrapper. The clause was unreachable, which is also the honest answer to the question I asked you earlier.

## What stayed

The two direct `AutoTokenizer` call sites in processor repair and bnb dequantization now route through `_load_mlx_tokenizer`, because they reach the same defect. `_remote_code_reference` stays, since the branch decision needs it.

`loader.py` goes from +242/-46 to +135/-14; the PR overall from 526 insertions to 294.

## Nothing is lost

The full version is commit **`d3ee66d0`**, the direct parent of the slimming commit and still in this branch's history. `git revert --no-commit 6e2f50e3` restores all of it. If you think the trust propagation or the refusal messaging is worth having, it is a good standalone PR on its own merits and I would review it -- it just should not ride along with a bugfix, because it makes the fix harder to reason about and to backport.

## Verification after slimming

209 tests pass in each of four environments: transformers 5.5.0 without torchvision, 5.5.0 with torchvision, 4.57.6, and against an installed `unsloth` 2026.8.7. Neighbouring MLX suites: 542 passed. The five families still load, the three control repos (gpt2, Llama-3.2, Qwen3) are byte-identical before and after, and the gpt2 special-token regression stays fixed.

**Comment by danielhanchen** [comment]
Re-validated the slimmed PR from scratch rather than assuming the earlier runs carried over. That found a bug I introduced when slimming, fixed in `e551764b`.

## What I broke

Cutting the trust plumbing took `kwargs.setdefault("trust_remote_code", False)` with it. That line is not part of that feature. transformers reads a missing `trust_remote_code` as `None`, and `resolve_trust_remote_code` answers `None` by calling `input()` behind a SIGALRM timeout. The blank `PretrainedConfig()` this fix injects forces `has_local_code` False, so a remote-code repo lands on exactly that branch.

Measured against `mlx-community/dbrx-instruct-4bit` on a real tty:

| | prompt shown | blocked |
|---|---|---|
| slim, no default (what I pushed in `6e2f50e3`) | `Do you wish to run the custom code? [y/N]` | **17.8s** |
| with the default restored | none | 2.7s, clean `ValueError` |

With stdin closed both end in the same `ValueError`, which is exactly why 210 passing tests and four environments did not catch it. It needed a pty.

## The first regression test I wrote was worthless

I patched `builtins.input` to raise `AssertionError`. It passed with the fix reverted. `resolve_trust_remote_code` wraps the prompt in `except Exception` and rewrites anything raised there into the same `ValueError` the passing path produces, so the assertion was swallowed and became the very error I was matching. The committed test records the call and returns `"n"` instead, and I verified it fails with the fix reverted and passes with it.

## Full re-validation of the slim PR

| | result |
|---|---|
| PR test files x 4 envs (tf 5.5 -/+ torchvision, tf 4.57.6, unsloth 2026.8.7) | 210 passed each |
| `tests/security` | 156 passed |
| CI behavioral gate (the 37-file list) | 1512 passed, 6 skipped |
| MLX export gates + neighbouring suites | 542 passed |
| edge cases (refusals, scope restore, 4x15 concurrent scopes, hub ids, malformed `auto_map`) | 32/32 |
| platform matrix | 8/8 |
| families | 5 fixed, 0 regressions, 3 controls identical |
| static integrity | no dangling refs to removed symbols; the one F401 is pre-existing on main |

The five families still load, gpt2 keeps its special tokens, and DBRX refuses identically whether `trust_remote_code` is omitted or passed as `False`.

## Still not covered

Apple Silicon and Windows CI are queued on a staging repo and have not reported. And nothing anywhere exercises this against real `mlx-lm` rather than the torch simulation shim, on any platform, because both test files install the shim unconditionally. That gap predates the PR and I have not closed it.

File: tests/test_mlx_save_export_regressions.py [hunk 1]
```diff
@@ -4232,3 +4233,276 @@ def sanitize(self, weights):
...
+def test_tokenizer_without_declared_class_keeps_class_default_specials(tmp_path):
+    """gpt2 and friends declare no tokenizer_class and ship no
+    special_tokens_map.json: bos/eos/unk come from the tokenizer class's
+    __init__ defaults. Loading the bare backend drops them, and an eos_token of
+    None leaves mlx-lm generation with no stop token."""
+    import json
+    import transformers
+    import unsloth_zoo.mlx.loader as loader
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained("openai-community/gpt2")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_mlx_save_export_regressions.py]: Confirmed and fixed in 06373ccf. Reproduced with an isolated HF cache: offline the test fails with "We couldn't connect to https://huggingface.co" before reaching _load_mlx_tokenizer, and online it pulls 2.8MB into the cache. This file is a hard gate in the Repo tests (CPU) lane, so that lands on anyone behind a proxy. The fixture is now built locally with tokenizers.BPE in the same shape as the published repo (serialized tokenizer.json, no tokenizer_class, no special_tokens_map.json, model_type gpt2). It also asserts against a bare PreTrainedTokenizerFast load of the same directory, which is what the branch used to return, so the assertion is tighter than the downloaded reference it replaced. Passes offline with an empty cache and still fails when the model_type lookup is reverted.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Staging CI results. The Apple Silicon job here is label-gated (`mlx-ci.yml:31-36`) and skips on fork PRs, so I replicated the branch onto a staging repo and labelled that copy instead, which runs the repo's real Apple Silicon workflow without touching the org queue.

## The repo's own CI, including Apple Silicon

| job | result | |
|---|---|---|
| **MLX install + import smoke (Apple Silicon, pinned)** | **pass** | 7m26s |
| Repo tests (CPU) | pass | 5m10s |
| Core drift (3 upstream pins) | pass | 6m8s |
| MLX (Linux CPU: numerics + suites) | pass | 5m43s |
| Python 3.10-3.13 collect | pass | 3m56s |
| Source lint | pass | 25s |
| Wheel build + import smoke | pass | 1m52s |

7/7. The Apple Silicon leg is the `pinned` matrix leg, so real `mlx==0.32.1` / `mlx-lm==0.31.3` / `mlx-vlm` on `macos-15`, not the simulation shim. That is the first real-hardware signal this branch has had.

It also settles both checks that were red here:

- **Core drift** was stale, not a defect: the staging branch is cut from fresh upstream `main`, and it goes green with no code change. A rerun could never have cleared it, because a rerun replays the merge commit as first resolved and the job executes the PR's own checked-out `tests/conftest.py`.
- **Repo tests (CPU)** passes with the `AutoImageProcessor` fix, in the no-torchvision configuration that produced the original failure.

## Cross-platform: my harness was broken, not this PR

The three per-OS jobs failed identically on ubuntu-latest and windows-latest: `200 failed, 80 passed`, every failure the same

```
ImportError: Please install Unsloth via `pip install unsloth`!
```

That is `unsloth_zoo/__init__.py:257`, which requires `find_spec("unsloth")`. The generated workflow installed `-e .` but not `unsloth`, so every test in the file errored at import, including ones this PR never touches. The repo's own `Repo tests (CPU)` lane installs it with `--no-deps "unsloth @ git+...@main" || true`; my generator did not. Fixed in the tooling with a regression assertion, and re-staged. No PR change was warranted, and none was made.

## Still outstanding

The re-staged per-OS jobs are queued; macOS-15 never reported on the previous attempt. I will post the per-OS table when they land. Worth stating plainly: those jobs run the PR's own test files, and both install the torch simulation shim unconditionally, so a green macOS-15 there proves portability, not real-MLX behaviour. The Apple Silicon pass above is the real-MLX evidence.
